README.md
# Zenodo upload

Uploads file to Zenodo.

[![npm version](https://badge.fury.io/js/%40iomeg%2Fzenodo-upload.svg)](https://badge.fury.io/js/%40iomeg%2Fzenodo-upload)
[![CI](https://github.com/iomega/zenodo-upload/workflows/CI/badge.svg)](https://github.com/iomega/zenodo-upload/actions?query=workflow%3ACI)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=iomega_zenodo-upload&metric=alert_status)](https://sonarcloud.io/dashboard?id=iomega_zenodo-upload)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=iomega_zenodo-upload&metric=coverage)](https://sonarcloud.io/dashboard?id=iomega_zenodo-upload)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3726851.svg)](https://doi.org/10.5281/zenodo.3726851)
[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3805/badge)](https://bestpractices.coreinfrastructure.org/projects/3805)
[![fair-software.eu](https://img.shields.io/badge/fair--software.eu-%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8B%20%20%E2%97%8F%20%20%E2%97%8F-yellow)](https://fair-software.eu)

A JavaScript library to create a new version of a [Zenodo](https://zenodo.org) upload with a file.

Makes a draft copy of an existing Zenodo upload.
After overwriting the file and version the upload is published.

Can be used to create a [DOI](https://doi.org) for a updated data file.
A Zenodo upload must already exist using this library.

[API documentation](https://iomega.github.io/zenodo-upload/modules/_index_.html#).

## Install

```shell
npm install @iomeg/zenodo-upload
```

## Usage

### As a command line tool

```shell
npx --package @iomeg/zenodo-upload zenodo_upload [--sandbox] [--no-checksum] <deposition_id> <file> <version> <access_token>
```

To create new version (`1.2.3`) of [https://zenodo.org/record/1234567](https://zenodo.org/record/1234567) by uploading a local file called `somefile`.

```shell
npx --package @iomeg/zenodo-upload zenodo_upload 1234567 somefile 1.2.3 sometoken
```

The `sometoken` string has to be replaced with a valid [Zenodo access token](https://sandbox.zenodo.org/account/settings/applications/tokens/new/).

### As a library

To create new version of [https://zenodo.org/record/1234567](https://zenodo.org/record/1234567).

Example usage using NodeJS:

```javascript
const fs = require('fs');
const zenodo_upload = require('@iomeg/zenodo-upload').default;

const deposition_id = 1234567;
const filename = 'somefile.txt';
fs.writeFileSync(filename, 'sometext', 'utf8');
const version = '1.2.3';
const access_token = 'sometoken';

zenodo_upload(deposition_id, filename, version, access_token)
    .then(r => console.log(`New zenodo entry ${r.doi} created`))
    .catch(e => console.error(e))
;
```

Or in modern javascript

```javascript
import fs from 'fs';
import zenodo_upload from '@iomeg/zenodo-upload';

const deposition_id = 1234567;
const filename = 'somefile.txt';
await fs.promises.writeFile(filename, 'sometext', 'utf8');
const version = '1.2.3';
const access_token = 'sometoken';

const result = await zenodo_upload(deposition_id, filename, version, access_token);
console.log(`New zenodo entry ${result.doi} created`);
```

To run the example code you will need a valid Zenodo access token and a deposition id that can be written to by that token.

## Development

To install dependencies:

```shell
yarn install
```

To run the project in development/watch mode. Your project will be rebuilt upon changes.

```shell
yarn start
```

To bundle the package to the dist folder.

```shell
yarn build
```

To runs the test watcher (Jest) in an interactive mode. By default, runs tests related to files changed since the last commit.

```shell
yarn test
```

To run linter and fix fixable errors.

```shell
yarn lint --fix
```

## Credits

This project was bootstrapped with [TSDX](https://github.com/jaredpalmer/tsdx).

This project follows the [fair-software-nl](https://fair-software.nl) recommendations.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.5.1] - 2021-05-10

### Fixed

* Dependabot alerts of hosted-git-info, lodash and handlebars

## [0.5.0] - 2020-04-02

### Fixed

* Publishing date not updated ([#13](https://github.com/iomega/zenodo-upload/issues/13))
* Unable to use concept DOI ([#12](https://github.com/iomega/zenodo-upload/issues/12))

## [0.4.1] - 2020-04-01

### Added

* Code examples in doc strings

### Fixed

* DraftDiscardedError and others not exported in entrypoint
* Vague error messages

## [0.4.0] - 2020-03-31

### Added

* Command line tool ([#1](https://github.com/iomega/zenodo-upload/issues/1))

### Fixed

* Example does not work ([#10](https://github.com/iomega/zenodo-upload/issues/10))

## [0.3.1] - 2020-03-27

### Fixed

* Error: Digest method not supported ([#7](https://github.com/iomega/zenodo-upload/issues/7))

## [0.3.0] - 2020-03-27

### Added

* Perform checksum check and discard on match ([#5](https://github.com/iomega/zenodo-upload/issues/5))

## [0.2.3] - 2020-03-25

### Removed

* Publishing of package to [https://github.com/iomega/zenodo-upload/packages](https://github.com/iomega/zenodo-upload/packages)

## [0.2.2] - 2020-03-25

### Fixes

* html url is undefined ([#2](https://github.com/iomega/zenodo-upload/issues/2))

## [0.2.1] - 2020-03-25

Tying to get automated publishing after a GitHub release is created to work.

## [0.2.0] - 2020-03-25

Initial release.

[Unreleased]: https://github.com//iomega/zenodo-upload/compare/v0.5.1...HEAD
[0.5.1]: https://github.com//iomega/zenodo-upload/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com//iomega/zenodo-upload/compare/v0.4.1...v0.5.0
[0.4.1]: https://github.com//iomega/zenodo-upload/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com//iomega/zenodo-upload/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com//iomega/zenodo-upload/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com//iomega/zenodo-upload/compare/v0.2.3...v0.3.0
[0.2.3]: https://github.com//iomega/zenodo-upload/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com//iomega/zenodo-upload/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com//iomega/zenodo-upload/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com//iomega/zenodo-upload/releases/tag/v0.2.0

test/cli.test.ts
import { main } from '../src/cli';
import { zenodo_upload } from '../src/zenodo_upload';
jest.mock('../src/zenodo_upload');

describe('main()', () => {
  let mockedExit: any;

  beforeAll(() => {
    mockedExit = jest.spyOn(process, 'exit').mockImplementation(exit => {
      throw Error(`${exit}`);
    });
  });

  describe('called with `--help`', () => {
    it('should show help', () => {
      expect.assertions(1);

      try {
        main(['--help']);
      } catch (error) {
        expect(mockedExit).toHaveBeenCalledWith(0);
      }
    });
  });

  describe('called with `1234 somefile 1.2.3 sometoken`', () => {
    beforeEach(() => {
      (zenodo_upload as jest.Mock).mockImplementation(() => {
        return {
          id: 12345,
          doi: 'somedio/12345',
          result: 'https://zenodo.org/record/12345',
        };
      });
    });

    it('should call zenodo_upload()', () => {
      expect.assertions(1);

      try {
        main(['1234', 'somefile', '1.2.3', 'sometoken']);
      } catch (error) {
        expect(zenodo_upload).toHaveBeenCalledWith(
          1234,
          'somefile',
          '1.2.3',
          'sometoken',
          { checksum: true, sandbox: false }
        );
      }
    });
  });
});


test/utils.test.ts
import { determine_api_base_url } from '../src/utils';

describe('determine_api_base_url()', () => {
  it.each([
    [false, 'https://zenodo.org/api'],
    [true, 'https://sandbox.zenodo.org/api'],
  ])('%b => %s', (arg, expected) => {
    expect(determine_api_base_url(arg)).toEqual(expected);
  });
});


test/create_draft.test.ts
import fetch from 'node-fetch';

import { ZenodoDraft } from '../src/zenodo_draft';
import { create_draft } from '../src/create_draft';

jest.mock('node-fetch');
const { Response } = jest.requireActual('node-fetch');

const mockedfetch = (fetch as any) as jest.Mock;

const mockedZenodoSandboxAPI = (url: string, init: RequestInit) => {
  if (
    url ===
      'https://sandbox.zenodo.org/api/deposit/depositions/518011/actions/newversion' &&
    init.method === 'POST'
  ) {
    return new Response(
      JSON.stringify({
        links: {
          latest_draft:
            'https://sandbox.zenodo.org/api/deposit/depositions/518042',
        },
      }),
      {
        status: 201,
        statusText: 'Created',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );
  } else if (
    url ===
      'https://sandbox.zenodo.org/api/deposit/depositions/512658/actions/newversion' &&
    init.method === 'POST'
  ) {
    return new Response(
      JSON.stringify({ status: 404, message: 'PID does not exist.' }),
      {
        status: 404,
        statusText: 'NOT FOUND',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );
  } else if (url === 'https://sandbox.zenodo.org/record/512658') {
    return new Response('', {
      status: 302,
      statusText: 'FOUND',
      headers: {
        Location: 'https://sandbox.zenodo.org/record/518011',
      },
    });
  }

  throw new Error('URL not mocked, ' + url);
};

const mockedZenodoAPI = (url: string, init: RequestInit) => {
  if (
    url ===
      'https://zenodo.org/api/deposit/depositions/518011/actions/newversion' &&
    init.method === 'POST'
  ) {
    return new Response(
      JSON.stringify({
        links: {
          latest_draft: 'https://zenodo.org/api/deposit/depositions/518042',
        },
      }),
      {
        status: 201,
        statusText: 'Created',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );
  } else if (
    url ===
      'https://zenodo.org/api/deposit/depositions/512658/actions/newversion' &&
    init.method === 'POST'
  ) {
    return new Response(
      JSON.stringify({ status: 404, message: 'PID does not exist.' }),
      {
        status: 404,
        statusText: 'NOT FOUND',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );
  } else if (url === 'https://zenodo.org/record/512658') {
    return new Response('', {
      status: 302,
      statusText: 'FOUND',
      headers: {
        Location: 'https://zenodo.org/record/518011',
      },
    });
  }

  throw new Error('URL not mocked, ' + url);
};

describe('create_draft()', () => {
  const access_token = 'sometoken';

  describe('with version DOI against sandbox Zenodo API', () => {
    let draft: ZenodoDraft;
    beforeAll(async () => {
      mockedfetch.mockImplementation(mockedZenodoSandboxAPI);
      draft = await create_draft(518011, access_token, { sandbox: true });
    });

    it('should create draft using version DOI', () => {
      const expected_url =
        'https://sandbox.zenodo.org/api/deposit/depositions/518011/actions/newversion';
      const expected_init = {
        method: 'POST',
        headers: {
          Authorization: 'Bearer sometoken',
        },
      };
      expect(fetch).toHaveBeenCalledWith(expected_url, expected_init);
    });

    it('should return a new draft', () => {
      expect(draft.toString()).toEqual(
        'https://sandbox.zenodo.org/api/deposit/depositions/518042'
      );
    });
  });

  describe('with version DOI against production Zenodo API', () => {
    let draft: ZenodoDraft;
    beforeAll(async () => {
      mockedfetch.mockImplementation(mockedZenodoAPI);
      draft = await create_draft(518011, access_token);
    });

    it('should create draft using version DOI', () => {
      const expected_url =
        'https://zenodo.org/api/deposit/depositions/518011/actions/newversion';
      const expected_init = {
        method: 'POST',
        headers: {
          Authorization: 'Bearer sometoken',
        },
      };
      expect(fetch).toHaveBeenCalledWith(expected_url, expected_init);
    });

    it('should return a new draft', () => {
      expect(draft.toString()).toEqual(
        'https://zenodo.org/api/deposit/depositions/518042'
      );
    });
  });

  describe('with wrong deposition id against sandbox Zenodo API', () => {
    it('should throw a not found error', async () => {
      expect.assertions(1);
      mockedfetch.mockImplementation((url: string) => {
        if (
          url ===
          'https://sandbox.zenodo.org/api/deposit/depositions/-1/actions/newversion'
        ) {
          return new Response(
            JSON.stringify({ status: 404, message: 'PID does not exist.' }),
            {
              status: 404,
              statusText: 'NOT FOUND',
              headers: {
                'Content-Type': 'application/json',
              },
            }
          );
        }
        return new Response('html page with not found error', {
          status: 404,
          statusText: 'NOT FOUND',
        });
      });
      try {
        await create_draft(-1, access_token, { sandbox: true });
      } catch (error) {
        expect(error).toEqual(
          new Error(
            'Zenodo communication error determining latest version from concept: NOT FOUND'
          )
        );
      }
    });
  });

  describe('with wrong deposition id against production Zenodo API', () => {
    it('should throw a not found error', async () => {
      expect.assertions(1);
      mockedfetch.mockImplementation((url: string) => {
        if (
          url ===
          'https://zenodo.org/api/deposit/depositions/-1/actions/newversion'
        ) {
          return new Response(
            JSON.stringify({ status: 404, message: 'PID does not exist.' }),
            {
              status: 404,
              statusText: 'NOT FOUND',
              headers: {
                'Content-Type': 'application/json',
              },
            }
          );
        }
        return new Response('html page with not found error', {
          status: 404,
          statusText: 'NOT FOUND',
        });
      });
      try {
        await create_draft(-1, access_token);
      } catch (error) {
        expect(error).toEqual(
          new Error(
            'Zenodo communication error determining latest version from concept: NOT FOUND'
          )
        );
      }
    });
  });

  describe('with wrong token against sandbox Zenodo API', () => {
    it('should throw an unauthorized error', async () => {
      expect.assertions(1);
      mockedfetch.mockImplementation(() => {
        return new Response(
          JSON.stringify({
            message: 'Wrong credentials',
            status: 401,
          }),
          {
            status: 401,
            statusText: 'UNAUTHORIZED',
            headers: {
              'Content-Type': 'application/json',
            },
          }
        );
      });
      try {
        await create_draft(518011, 'badtoken', { sandbox: true });
      } catch (error) {
        expect(error).toEqual(
          new Error(
            'Zenodo API communication error creating draft: UNAUTHORIZED, {"message":"Wrong credentials","status":401}'
          )
        );
      }
    });
  });

  describe('with concept DOI against sandbox Zenodo API', () => {
    let draft: ZenodoDraft;
    beforeAll(async () => {
      mockedfetch.mockImplementation(mockedZenodoSandboxAPI);
      draft = await create_draft(512658, access_token, { sandbox: true });
    });

    it('should try to create draft using concept DOI', () => {
      expect(fetch).toHaveBeenCalledWith(
        'https://sandbox.zenodo.org/api/deposit/depositions/512658/actions/newversion',
        {
          method: 'POST',
          headers: {
            Authorization: 'Bearer sometoken',
          },
        }
      );
    });

    it('should fetch concept DOI page', () => {
      expect(fetch).toHaveBeenCalledWith(
        'https://sandbox.zenodo.org/record/512658',
        {
          redirect: 'manual',
          follow: 0,
        }
      );
    });

    it('should create draft using latest version DOI', () => {
      const expected_url =
        'https://sandbox.zenodo.org/api/deposit/depositions/518011/actions/newversion';
      const expected_init = {
        method: 'POST',
        headers: {
          Authorization: 'Bearer sometoken',
        },
      };
      expect(fetch).toHaveBeenCalledWith(expected_url, expected_init);
    });

    it('should return a new draft', () => {
      expect(draft.toString()).toEqual(
        'https://sandbox.zenodo.org/api/deposit/depositions/518042'
      );
    });
  });
});


src/cli.ts
import yargs from 'yargs';
import { zenodo_upload } from './zenodo_upload';

function builder(instance: any) {
  return instance
    .positional('deposition_id', {
      type: 'number',
      description:
        'Zenodo deposition identifier. For eg. use 3726935 as identifier for https://zenodo.org/record/3726935.',
    })
    .positional('file', {
      type: 'string',
      description: 'Path to file to upload to Zenodo',
    })
    .positional('version', {
      type: 'string',
      description: 'Version string to use in new Zenodo upload',
    })
    .positional('access_token', {
      type: 'string',
      description:
        'Zenodo access token. During token generation check the `deposit:actions` and `deposit:write` scopes. Generate at https://zenodo.org/account/settings/applications/tokens/new/',
    })
    .option('sandbox', {
      type: 'boolean',
      default: false,
      description:
        'Publish to Zenodo sandbox (https://sandbox.zenodo.org) environment instead of Zenodo production environment',
    })
    .option('checksum', {
      type: 'boolean',
      default: true,
      description:
        'Only create new Zenodo version when checksum of file is different',
    });
}

function handler(argv: any) {
  zenodo_upload(
    argv.deposition_id,
    argv.file,
    argv.version,
    argv.access_token,
    { sandbox: argv.sandbox, checksum: argv.checksum }
  )
    .then(result => {
      console.log(
        `Generated new versioned DOI: ${result.doi}, can take a while to be found`
      );
      console.log(`Generated new Zenodo upload: ${result.html}`);
      process.exit(0);
    })
    .catch(e => {
      console.error(e.message);
      process.exit(1);
    });
}

/**
 * Command line tool implementation
 *
 * @param args When empty will use arguments from process.argv
 */
export function main(args: string[] = []) {
  const parser = yargs
    .version(false) // disabled interferes with <version> positional argument
    .usage(
      '$0 [--sandbox] [--no-checksum] <deposition_id> <file> <version> <access_token>',
      'Create new version of a Zenodo upload',
      builder
    );
  let argv;
  if (args.length > 0) {
    argv = parser.parse(args);
  } else {
    argv = parser.parse();
  }
  handler(argv);
}


src/utils.ts
import { basename } from 'path';
import { fromFile } from 'hasha';

export function determine_api_base_url(sandbox: boolean) {
  if (sandbox) {
    return 'https://sandbox.zenodo.org/api';
  }
  return 'https://zenodo.org/api';
}

export function auth_headers(access_token: string) {
  return {
    Authorization: `Bearer ${access_token}`,
  };
}

export interface DepositionFile {
  id: string;
  filename: string;
  filesize: number;
  checksum: string;
}

export class FilePresentError extends Error {
  constructor(m: string) {
    super(m);
    Object.setPrototypeOf(this, FilePresentError.prototype);
  }
}

export async function is_same_file_present(
  file: string,
  bag: DepositionFile[]
) {
  const filename = basename(file);
  const other = bag.find(f => f.filename === filename);
  if (!other) {
    return false;
  }
  const checksum = await fromFile(file, { algorithm: 'md5' });
  if (checksum === other.checksum) {
    throw new FilePresentError(`File ${file} is present`);
  }
  return true;
}


src/zenodo_draft.ts
import fs from 'fs';
import { basename } from 'path';
import fetch from 'node-fetch';
import { fromFile as fileTypefromFile } from 'file-type';
import { auth_headers, DepositionFile, is_same_file_present } from './utils';

export interface PublishResult {
  id: number;
  doi: string;
  html: string;
}

/**
 * Manage draft entry on Zenodo
 *
 * Example of managing a already created Zenodo draft.
 *
 * ```javascript
 * import { ZenodoDraft } from '@iomeg/zenodo-upload';
 *
 * const draft_url = 'https://zenodo.org/api/deposit/depositions/1234567';
 * const access_token = 'sometoken';
 * const checksum = true;
 * const draft  = new ZenodoDraft(draft_url, access_token, checksum);
 * await draft.add_file('./somefile');
 * await draft.set_version('1.2.3');
 * await draft.publish();
 * ```
 *
 * A draft can be created on the Zenodo website by creating a new version without publishing it or with the {@link create_draft} function.
 */
export class ZenodoDraft {
  private url: string;
  private access_token: string;
  private checksum: boolean;
  private bucket = '';
  private _metadata: any;
  private files: DepositionFile[] = [];

  /**
   *
   * @param url URL of draft Zenodo upload, something like https://zenodo.org/api/deposit/depositions/1234567
   * @param access_token The [Zenodo personal access token](https://sandbox.zenodo.org/account/settings/applications/tokens/new/) with the `deposit:actions` and `deposit:write` scopes.
   * @param checksum When true then an added file is checksum matched with file already in draft. If it matches an error is thrown.
   */
  constructor(url: string, access_token: string, checksum: boolean) {
    this.url = url;
    this.access_token = access_token;
    this.checksum = checksum;
  }

  /**
   * Set version of draft
   *
   * For example
   *
   * ```javascript
   * await draft.set_version('1.2.3');
   * ```
   *
   * @param version Update draft with given version
   * @param publication_date Date of publication. Defaults to today.
   *
   * @throws {Error} When communication with Zenodo API fails
   */
  async set_version(version: string, publication_date = new Date()) {
    const metadata = await this.get_metadata();
    metadata.version = version;
    metadata.publication_date = publication_date.toISOString().slice(0, 10);
    await this.set_metadata(metadata);
  }

  /**
   * Get copy of the metadata.
   *
   * @returns metadata object of draft
   *
   * @throws {Error} When communication with Zenodo API fails
   */
  async get_metadata() {
    if (!this._metadata) {
      await this.build_cache();
    }
    return JSON.parse(JSON.stringify(this._metadata));
  }

  /**
   * Overwrite complete metadata of draft
   *
   * For example
   *
   * ```javascript
   * const metadata = await draft.get_metadata();
   * metadata.title = 'My new title';
   * await draft.set_metadata(metadata);
   * ```
   *
   * @param metadata New metadata object for draft
   *
   * @throws {Error} When communication with Zenodo API fails
   */
  async set_metadata(metadata: any) {
    const body = JSON.stringify({ metadata });
    const init = {
      method: 'PUT',
      headers: {
        ...auth_headers(this.access_token),
        'Content-Type': 'application/json',
      },
      body,
    };
    const response = await fetch(this.url, init as any);
    if (response.ok) {
      const json = await response.json();
      this.fill_cache(json);
    } else {
      throw new Error(
        `Zenodo API communication error updating metadata: ${
          response.statusText
        }, ${await response.text()}`
      );
    }
  }

  /**
   * Add file to draft
   *
   * For example to add a file called `./somefile`.
   *
   * ```javascript
   * await draft.add_file('./somefile');
   * ```
   *
   * @param file Path to file
   *
   * @throws {FilePresentError} When checksum=true and file checksum matched with file already in draft.
   * @throws {Error} When communication with Zenodo API fails
   */
  async add_file(file: string) {
    if (!this.bucket) {
      await this.build_cache();
    }
    const stat = await fs.promises.stat(file);
    const content_type = await fileTypefromFile(file);
    if (this.checksum) {
      await is_same_file_present(file, this.files);
    }

    const body = fs.createReadStream(file);
    const init = {
      method: 'PUT',
      headers: {
        ...auth_headers(this.access_token),
        'Content-Type': content_type,
        'Content-Length': stat.size,
      },
      body,
    };
    const filename = basename(file);
    const url = `${this.bucket}/${filename}`;
    const response = await fetch(url, init as any);
    if (response.ok) {
      const rbody: DepositionFile = await response.json();
      this.add_file2cache(rbody);
      return rbody;
    } else {
      throw new Error(
        `Zenodo API communication error adding file: ${
          response.statusText
        }, ${await response.text()}`
      );
    }
  }

  /**
   * Publish draft Zenodo upload
   *
   * For example
   *
   * ```javascript
   * await draft.publish();
   * ```
   *
   * @throws {Error} When communication with Zenodo API fails
   */
  async publish() {
    const url = this.url + '/actions/publish';
    const init = {
      method: 'POST',
      headers: auth_headers(this.access_token),
    };
    const response = await fetch(url, init as any);
    if (response.ok) {
      const body = await response.json();
      this.fill_cache(body);

      const result: PublishResult = {
        id: body.id,
        doi: body.links.doi,
        html: body.links.latest_html,
      };
      return result;
    } else {
      throw new Error(
        `Zenodo API communication error publishing: ${
          response.statusText
        }, ${await response.text()}`
      );
    }
  }

  /**
   * Discard the draft.
   *
   * Any other methods on the object will fail after discarding.
   *
   * For example
   *
   * ```javascript
   * await draft.discard();
   * ```
   * @throws {Error} When communication with Zenodo API fails
   */
  async discard() {
    const init = {
      method: 'DELETE',
      headers: auth_headers(this.access_token),
    };
    const response = await fetch(this.url, init as any);
    if (!response.ok) {
      throw new Error(
        `Zenodo API communication error discarding: ${
          response.statusText
        }, ${await response.text()}`
      );
    }
  }

  toString() {
    return this.url;
  }

  private add_file2cache(filedep: DepositionFile) {
    const index = this.files.findIndex(f => f.id === filedep.id);
    if (index === -1) {
      this.files.push(filedep);
    } else {
      this.files[index] = filedep;
    }
  }

  private fill_cache(record: any) {
    this.bucket = record.links.bucket;
    this._metadata = record.metadata;
    this.files = record.files;
  }

  private async build_cache() {
    const init = {
      method: 'GET',
      headers: auth_headers(this.access_token),
    };
    const response = await fetch(this.url, init);
    if (response.ok) {
      const body = await response.json();
      this.fill_cache(body);
    } else {
      throw new Error(
        `Zenodo API communication error fetching draft: ${
          response.statusText
        }, ${await response.text()}`
      );
    }
  }
}


src/zenodo_upload.ts
import { create_draft, Options, DEFAULT_OPTIONS } from './create_draft';
import { FilePresentError } from './utils';

export class DraftDiscardedError extends Error {
  constructor(m: string) {
    super(m);
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, DraftDiscardedError.prototype);
  }
}

/**
 * Create a new version of a Zenodo upload with a file.
 *
 * Makes a draft copy of an existing Zenodo upload.
 * After overwriting the file and version the upload is published.
 *
 * For example
 *
 * ```javascript
 * import fs from 'fs';
 * import zenodo_upload from '@iomeg/zenodo-upload';
 *
 * const deposition_id = 1234567;
 * const filename = 'somefile.txt';
 * await fs.promises.writeFile(filename, 'sometext', 'utf8');
 * const version = '1.2.3';
 * const access_token = 'sometoken';
 *
 * const result = await zenodo_upload(deposition_id, filename, version, access_token);
 * console.log(`New zenodo entry ${result.doi} created`);
 * ```
 *
 * @param deposition_id Zenodo upload identifier
 * @param file Path to file to upload
 * @param version Version to use
 * @param access_token The [Zenodo personal access token](https://sandbox.zenodo.org/account/settings/applications/tokens/new/) with the `deposit:actions` and `deposit:write` scopes.
 * @param options Advanced options
 *
 * @throws {DraftDiscardedError} When `options.checksum` and `file` is already present.
 *
 * @returns The id, url of the html page and the DOI of the new version.
 */
export async function zenodo_upload(
  deposition_id: number,
  file: string,
  version: string,
  access_token: string,
  options: Options = DEFAULT_OPTIONS
) {
  const { sandbox = false, checksum = false } = options;
  const draft = await create_draft(deposition_id, access_token, {
    sandbox,
    checksum,
  });
  try {
    await draft.add_file(file);
  } catch (error) {
    if (checksum && error instanceof FilePresentError) {
      await draft.discard();
      throw new DraftDiscardedError(
        `Zenodo draft version ${draft} has been discarded`
      );
    }
    throw error;
  }
  await draft.set_version(version);
  return await draft.publish();
}


src/create_draft.ts
import fetch from 'node-fetch';
import { auth_headers, determine_api_base_url } from './utils';
import { ZenodoDraft } from './zenodo_draft';

export interface Options {
  /**
   * Whether to use the production Zenodo environment or the sandbox environment. False by default.
   */
  sandbox?: boolean;
  /**
   * If true will check if MD5 checksum of `file` is already present in previous Zenodo upload. False by default.
   */
  checksum?: boolean;
}

export const DEFAULT_OPTIONS = Object.freeze({
  sandbox: false,
  checksum: false,
});

function determine_base_url(sandbox: boolean) {
  if (sandbox) {
    return 'https://sandbox.zenodo.org/record/';
  }
  return 'https://zenodo.org/record/';
}

function deposition_id_from_pageurl(url: string) {
  return parseInt(url.slice(url.lastIndexOf('/') + 1));
}

function newversion_url(sandbox: boolean, deposition_id: number) {
  const api_base_url = determine_api_base_url(sandbox);
  return (
    api_base_url + `/deposit/depositions/${deposition_id}/actions/newversion`
  );
}

async function resolve_concept_to_latestversion(
  sandbox: boolean,
  deposition_id: number
) {
  const url = determine_base_url(sandbox) + deposition_id;
  const response = await fetch(url, { redirect: 'manual', follow: 0 });
  if (response.status === 302) {
    const latestversion_pageurl = response.headers.get('Location');
    return deposition_id_from_pageurl(latestversion_pageurl!);
  }
  throw new Error(
    `Zenodo communication error determining latest version from concept: ${response.statusText}`
  );
}

/**
 * Create a draft from an existing Zenodo upload
 *
 * For example to create a draft from [https://zenodo.org/record/1234567](https://zenodo.org/record/1234567).
 *
 * ```javascript
 * import { create_draft } from '@iomeg/zenodo-upload';
 *
 * const deposition_id = 1234567;
 * const access_token = 'sometoken';
 *
 * const draft = await create_draft(deposition_id, access_token);
 * draft.discard();
 * ```
 *
 * @param deposition_id The deposition identifier of a Zenodo upload
 * @param access_token The [Zenodo personal access token](https://sandbox.zenodo.org/account/settings/applications/tokens/new/) with the `deposit:actions` and `deposit:write` scopes.
 * @param options
 *
 * @throws {Error} When communication with Zenodo API fails
 */
export async function create_draft(
  deposition_id: number,
  access_token: string,
  options: Options = DEFAULT_OPTIONS
): Promise<ZenodoDraft> {
  const { sandbox = false, checksum = false } = options;
  const url = newversion_url(sandbox, deposition_id);
  const init = {
    method: 'POST',
    headers: auth_headers(access_token),
  };
  const response = await fetch(url, init);
  if (response.ok) {
    const body = await response.json();
    return new ZenodoDraft(body.links.latest_draft, access_token, checksum);
  } else {
    if (response.status === 404) {
      // Possibly a concept DOI, will try to get latest version DOI and try again
      const latestversion_deposition_id = await resolve_concept_to_latestversion(
        sandbox,
        deposition_id
      );
      return await create_draft(
        latestversion_deposition_id,
        access_token,
        options
      );
    }
    throw new Error(
      `Zenodo API communication error creating draft: ${
        response.statusText
      }, ${await response.text()}`
    );
  }
}
